gpflow/actions.py

```diff
@@ -0,0 +1,386 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import collections
+import inspect
+import itertools
+from timeit import default_timer as timer
+from typing import Any, Callable, Dict, Optional, Sequence, Union
+
+import numpy as np
+import tensorflow as tf
+
+from . import get_default_session
+from .models import Model
+
+
+class Watcher:
+    """
+    Watcher is a sort of time manager.
+    
+    It holds information when timer is started, and when it is stopped.
+    If elapsed time is asked and timer is not stopped, then different between
+    current time and started is returned. Otherwise, different between
+    start and stop time.
+    """
+
+    def __init__(self):
+        self._start = None
+        self._stop = None
+
+    def start(self):
+        """
+        Set start time [µs].
+        """
+        self._start = timer()
+    
+    def stop(self):
+        """
+        Set stop time [µs].
+        """
+        self._stop = timer()
+    
+    @property
+    def elapsed(self):
+        """
+        Elapsed time [µs] between start and stop timestamps. If stop is empty then
+        returned time is difference between start and current timestamp.
+        """
+        if self._stop is None:
+            return timer() - self._start
+        return self._stop - self._start
+    
+
+class ActionContext:
+    """
+    Action context contains sharable data between sequence of action runs.
+
+        :param owner: Action to which context originally belongs to.
+        :param session: Tensorflow session which will be used when necessary by
+            all dependent actions ran by owner.
+    """
+
+    def __init__(self, owner: 'Action', session: Optional[tf.Session] = None):
+        if session is None:
+            session = tf.get_default_session()
+            session = get_default_session() if session is None else session
+        self.session = session
+        self.owner = owner
+    
+    @property
+    def iteration(self) -> int:
+        """
+        Current iteration number.
+        In fact, this is proxy method for Loop's owner interation.
+
+            :return: Iteration number or None, in case when owner doesn't
+                have any references to iteration.
+        """
+        return _get_attr(self.owner, iteration=None)
+    
+    @property
+    def time_spent(self) -> float:
+        """
+        Elapsed time since owner started running dependent actions.
+
+            :return: Spent time [µs] by owner on running up to the call.
+        """
+        return self.owner.watcher.elapsed
+
+
+class Action(metaclass=abc.ABCMeta):
+    """
```

Comment:
These actions seem to play the role of "OptimizationAction"s, specifically - or it is fine to keep this very general terminology?

Reply:
Later we can add TensorboardAction, LoggingAction and cetera.

Reply:
But these are still actions done during the optimization?

Reply:
Or anything else :)
Optimization is only part of this process, action is more generalized action. E.g. loop is not optimization, condition too. Logger action is not part of optimization.

Reply:
The point is that action can be anything (including optimization). Technically, yes, loop created to help to simplify complex optimization steps, but still is not correct to call it "OptimiztionLoop", as you can run w/o actual optimization.